src/rust/engine/process_execution/src/bounded.rs

```diff
@@ -0,0 +1,403 @@
+}
+
+///
+/// A CommandRunner wrapper which limits the number of concurrent requests and which provides
+/// concurrency information to the process being executed.
+///
+#[derive(Clone)]
+pub struct CommandRunner {
+  inner: Arc<dyn crate::CommandRunner>,
+  sema: AsyncSemaphore,
+}
+
+impl CommandRunner {
+  pub fn new(
+    executor: &Executor,
+    inner: Box<dyn crate::CommandRunner>,
+    bound: usize,
+  ) -> CommandRunner {
+    CommandRunner {
+      inner: inner.into(),
+      sema: AsyncSemaphore::new(
+        executor,
+        bound,
+        // TODO: Make configurable.
+        Duration::from_millis(200),
+      ),
+    }
+  }
+}
+
+#[async_trait]
+impl crate::CommandRunner for CommandRunner {
+  async fn run(
+    &self,
+    context: Context,
+    workunit: &mut RunningWorkunit,
+    process: Process,
+  ) -> Result<FallibleProcessResultWithPlatform, String> {
+    let semaphore_acquisition = self.sema.acquire(process.concurrency_available);
+    let permit = in_workunit!(
+      context.workunit_store.clone(),
+      "acquire_command_runner_slot".to_owned(),
+      WorkunitMetadata {
+        level: Level::Trace,
+        ..WorkunitMetadata::default()
+      },
+      |workunit| async move {
+        let _blocking_token = workunit.blocking();
+        semaphore_acquisition.await
+      }
+    )
+    .await;
+
+    loop {
+      let mut process = process.clone();
+      let concurrency_available = permit.concurrency();
+      log::debug!(
+        "Running {} under semaphore with concurrency id: {}, and concurrency: {}",
+        process.description,
+        permit.concurrency_slot(),
+        concurrency_available,
+      );
+
+      // TODO: Both of these templating cases should be implemented at the lowest possible level:
+      // they might currently be applied above a cache.
+      if let Some(ref execution_slot_env_var) = process.execution_slot_variable {
+        process.env.insert(
+          execution_slot_env_var.clone(),
+          format!("{}", permit.concurrency_slot()),
+        );
+      }
+      if process.concurrency_available > 0 {
+        let concurrency = format!("{}", permit.concurrency());
+        let mut matched = false;
+        process.argv = std::mem::take(&mut process.argv)
+          .into_iter()
+          .map(
+            |arg| match CONCURRENCY_TEMPLATE_RE.replace_all(&arg, &concurrency) {
+              Cow::Owned(altered) => {
+                matched = true;
+                altered
+              }
+              Cow::Borrowed(_original) => arg,
+            },
+          )
+          .collect();
+        if !matched {
+          return Err(format!("Process {} set `concurrency_available={}`, but did not include the `{}` template variable in its arguments.", process.description, process.concurrency_available, *CONCURRENCY_TEMPLATE_RE));
+        }
+      }
+
+      let running_process = self.inner.run(context.clone(), workunit, process.clone());
```

Comment:
I'm trying to understand why it's safe to repeatedly mutate the `Process` inside `loop`. Naively, I would think that would cause the `Process` to restart because it's a new argv -- but iiuc we're only trying to preempt/pause it, not restart.

Reply:
The mutation that is happening here is of a copy of the `Process` struct that is created on line 77: so each iteration of the loop effectively gets a fresh copy of the original argument to this function.

> Naively, I would think that would cause the Process to restart because it's a new argv -- but iiuc we're only trying to preempt/pause it, not restart.

So, the terminology might be confusing here: the effect of this loop is that we will _restart_ the process if `permit.notified_concurrency_changed()`. We cannot affect a process once it is already running, and it's why we have the preemption threshold: restarting a process once it has made a bunch of progress isn't cheap, but if it has just barely started, we can be fairly sure we're not losing much progress when we restart it.

So [preemption](https://en.wikipedia.org/wiki/Preemption_(computing))...
> "temporarily interrupting an executing task, with the intention of resuming it at a later time"

...is accurate, but misleading. Because no operating system that I know of speculatively restarts tasks like this. I'm just not aware of a better word for it.

Reply:
Oh, correction: kubernetes apparently uses preemption this way too, in that it will kill and reschedule pods to make room for other pods: https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption

Reply:
Oh, okay! That would be helpful to add as a comment, including an explicit callout that we implement pre-emption differently than the reader may expect with that term.